### PR TITLE
Issue 6516: Add metric for Table Segment Credits

### LINKED ISSUE #6516: Add metric for Table Segment Credits
**Is your feature request related to a problem? Please describe.**
Add a new metric to expose the available credits for Table Segment credits.

**Describe the solution you'd like**
Add a new metric to expose the available credits for Table Segment credits.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
Added a new metric to keep track of Table Segment used credits.

**Purpose of the change**  
Fixes #6516.

**What the code does**  
- Adds a new (gauge) metric for the credits of (hash-based) Table Segments called: `TABLE_SEGMENT_USED_CREDITS = PREFIX + "segmentstore.tablesegment.used_credits"`
- The metric gets reported in `ContainerKeyIndex` upon a new update to a Table Segment. This allows us to report the metric even if there are no credit releases.
- Tested the metric locally and it is being reported.

**How to verify it**  
All tests are working as before. Tested the metric locally and observed that it is being reported:
Screenshot from 2022-01-10 17-25-53

### CODE REVIEW

**Comment by abhinb** [comment]
would this graph be available by default in SS dashboard?

**Review by abhinb — APPROVED** [review]

**Comment by RaulGracia** [comment]
@sachin-j-joshi @abhinb I have done a small change to get the right (hash-based table segment) segment tags when reporting the metric. I have deployed it locally again and it works. Please, have another look when you have the time.

**Comment by sachin-j-joshi** [comment]
@RaulGracia I am not sure whether reporting per segment metric is a scalable or not. Not that I'm against the idea per se, but more like should we be concerned by the amount of metric data generated. (how many table segments we are talking about here?)

**Comment by RaulGracia** [comment]
@sachin-j-joshi we are already reporting metrics adding per-segment tags:
https://github.com/pravega/pravega/blob/9aaa35cad994e36c7605bffe763874b5bc5e87f8/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/SegmentStatsRecorderImpl.java

As far as I know, I have done experiments with thousands of Segments and I have seen no issues at the moment regarding the way Pravega publishes these metrics adding per-segment tags. Also note that metrics are accumulated locally and then reported every `metrics.output.frequency.seconds` seconds (i.e., doing an operation on a segment does not imply that we are immediately reporting that to the metrics).

More importantly, this metric is only published for hash-based table segments (fixed-length table segments do not have credits). This means that the only table segments of this kind in the system are the ones that Pravega uses to store metadata (i.e., container/storage metadata x num. containers, plus few more related to Controller metadata). I don't think that this can have any significant impact on the system, given that we are currently publishing +200 metrics (adding tags on a per segment, per host, or per container, depending on the case).

**Comment by RaulGracia** [comment]
@sachin-j-joshi will run system tests before merging it, just to be on the safe side.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java [hunk 1]
```diff
@@ -528,4 +529,20 @@ public static void outstandingEventProcessorBytes(String processorName, int cont
     }
 
     //endregion
+
+    //region ContainerKeyIndex
+
+    /**
+     * Reports the number of credits used for a given Hash-Based Table Segment. Note that these Segments are exclusively
+     * used for internal metadata storage and may have a special name format. For this reason, we report the Segment name
```
Comment (abhinb) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java:537]: Are we going to see full name starting from the underscore .."_system/<stream>/<segmnt>" with this?

Comment (RaulGracia) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java:537]: Yes, that's correct. If you see the first screenshot I posted, we also see the segment name like `_system/containers/storage_metadata_0` (using the `segmentTags` method). But other tags are added apart from that one, I think. Due to the different naming that these internal segments may have (also in tests), I opted to just report the credits and the segment name, so we are sure that parsing the tags do not lead into issues.

**Comment by RaulGracia** [comment]
System tests have passed: `test-pravega-continuous/4322` `test-pravega-continuous-1/1918`. cc/ @abhinb @sachin-j-joshi

**Review by sachin-j-joshi — APPROVED** [review]
LGTM

**Pull request merged**